cruise-control-core/src/main/java/com/linkedin/cruisecontrol/monitor/sampling/aggregator/RawMetricValues.java

```diff
@@ -49,20 +50,69 @@ protected int length() {
    *
    * @param numWindowsToKeep the total number of windows to keep track of.
    * @param minSamplesPerWindow the minimum required samples for a window to not involve any {@link Extrapolation}.
+   * @param numMetricTypesInSample the total number of raw metric types stored by {@link #_windowValuesByMetricId}
    */
-  public RawMetricValues(int numWindowsToKeep, int minSamplesPerWindow) {
+  public RawMetricValues(int numWindowsToKeep, byte minSamplesPerWindow, int numMetricTypesInSample) {
     if (numWindowsToKeep <= 1) {
       throw new IllegalArgumentException("The number of windows should be at least 2 because at least one available"
                                              + " window and one current window are needed.");
     }
-    _valuesByMetricId = new HashMap<>();
-    _counts = new short[numWindowsToKeep];
+    _windowValuesByMetricId = new HashMap<>(numMetricTypesInSample);
+    _counts = new byte[numWindowsToKeep];
     _extrapolations = new BitSet(numWindowsToKeep);
     _validity = new BitSet(numWindowsToKeep);
     _minSamplesPerWindow = minSamplesPerWindow;
+    _halfMinRequiredSamples = (byte) Math.max(1, _minSamplesPerWindow / 2);
     _oldestWindowIndex = Long.MAX_VALUE;
   }
 
+  /**
+   * Update {@link #_validity} and {@link #_extrapolations} for the previous and next array indices of the given arrayIndex.
+   * @param arrayIndex Array index.
+   */
+  private void maybeUpdateValidityAndExtrapolationOfPrevAndNextFor(int arrayIndex) {
+    if (_counts[arrayIndex] >= _minSamplesPerWindow) {
+      _extrapolations.clear(arrayIndex);
```

Comment:
This seems redundant since once the sample count reaches `_minSamplesPerWindow`, it is already [cleared from _extrapolations](https://github.com/linkedin/cruise-control/blob/ca1f91384c0debb422f8b59a52eb52cac0d42685/cruise-control-core/src/main/java/com/linkedin/cruisecontrol/monitor/sampling/aggregator/RawMetricValues.java#L414)